src/beliefs/discrete.jl

```diff
@@ -1,133 +1,107 @@
-"""
-mutable struct DiscreteBelief
+# needs pomdp for state_index in pdf(b, s)
+# needs list of ordered_states for rand(b)
+mutable struct DiscreteBelief{P<:POMDP, S}
+    pomdp::P
+    state_list::Vector{S}       # vector of ordered states
     b::Vector{Float64}
 end
+function DiscreteBelief(pomdp)
+    state_list = ordered_states(pomdp)
+    ns = n_states(pomdp)
 
-# Constructor with uniform belief
-DiscreteBelief(n::Int64) = DiscreteBelief(zeros(n) + 1.0/n)
+    b = ones(ns) / ns
 
-mutable struct DiscreteUpdater{P<:POMDP} <: Updater
-    pomdp::P
+    return DiscreteBelief(pomdp, state_list, b)
 end
 
-vec(b::DiscreteBelief) = b.b
 
-Base.length(b::DiscreteBelief) = length(b.b)
-index(b::DiscreteBelief, i::Int64) = i
-weight(b::DiscreteBelief, i::Int64) = b.b[i]
-iterator(b::DiscreteBelief) = filter(i->b[i]>0.0, 1:length(b))
-rand(rng::AbstractRNG, b::DiscreteBelief) = sample(rng, Weights(b.b)) # This will return an integer - seems like it should actually return an object of the state type of the problem
-pdf(b::DiscreteBelief, s::Int) = b.b[s]/sum(b.b) # only works when the state type is integer
+pdf(b::DiscreteBelief, s) = b.b[state_index(b.pomdp, s)]
+
+function rand(rng::AbstractRNG, b::DiscreteBelief)
+    i = sample(rng, Weights(b.b))
+    return b.state_list[i]
+end
 
 function Base.fill!(b::DiscreteBelief, x::Float64)
     fill!(b.b, x)
     return b
 end
 
-function Base.fill!(b::DiscreteBelief, idxs::Vector{Int64}, vals::Vector{Float64})
-    fill!(b.b, 0.0)
-    for i = 1:length(idxs)
-        index = idxs[i]
-        index > 0 ? (b[index] = vals[i]) : nothing 
+# should the belief just store the length so we don't have to count?
+Base.length(b::DiscreteBelief) = length(b.b)
+
+# I think this is a disaster and should be removed, but SARSOP.jl uses it
+# 
+# alphas are |A|x|S| (LD: it looks like |S|x|A|)
+# computes dot product of alpha vectors and belief
+# util is array with utility of each alpha vecotr for belief b
+function product(alphas::Matrix{Float64}, b::DiscreteBelief)
+    @assert size(alphas, 1) == length(b) "Alpha and belief sizes not equal"
+    n = size(alphas, 2) 
+    util = zeros(n)
+    for i = 1:n
+        s = 0.0
+        for j = 1:length(b)
+            s += alphas[j,i]*b[j]
+        end
+        util[i] = s
     end
-    b
+    return util
 end
 
-function Base.setindex!(b::DiscreteBelief, x::Float64, i::Int64) 
-    b.b[i] = x
-    b
-end
 
-Base.getindex(b::DiscreteBelief, i::Int64) = b.b[i]
 
-function Base.copy!(b1::DiscreteBelief, b2::DiscreteBelief)
-    copy!(b1.b, b2.b)
-    return b1
+mutable struct DiscreteUpdater{P<:POMDP} <: Updater
+    pomdp::P
 end
 
-Base.sum(b::DiscreteBelief) = sum(b.b)
-
-create_belief(bu::DiscreteUpdater) = DiscreteBelief(n_states(bu.pomdp))
+# Will this cause excessive calls to DiscreteBelief ?
+create_belief(bu::DiscreteUpdater) = DiscreteBelief(bu.pomdp)
```

Comment:
I think we should get rid of `create_belief` It is left over from the earliest days of POMDPs. Instead someone should use `initialize_belief` or the `uniform_belief` I mentioned above.

If you think it might be used elsewhere, you can make it throw a deprecation warning, (I think you can use the @deprecated macro for this).

Reply:
I am totally OK with this. I had no idea why this was there; I left it in because the old implementation had it.